### PR TITLE
Implementation of the slip boundary condition.

### PR DESCRIPTION
**Pull request by iyer-arvind**

### CODE REVIEW

**Comment by FreddieWitherden** [comment]
Until we have another BC using the 'slp-ghost' state move the macro into the same file as the BC and set `cflux_state = None` in the class.

**Comment by iyer-arvind** [comment]
Freddie, there are a few lines like lines 31, 32, 33, 52, 56, 57, 58, 80, 81 which have a extra space in the indentation. You may wat to fix those. before merge.

**Comment by FreddieWitherden** [comment]
Fixed in the current head of the branch.

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%include file='pyfr.solvers.euler.kernels.rsolvers.${rsolver}'/>
+<%include file='pyfr.solvers.navstokes.kernels.bcs.common'/>
+<%include file='pyfr.solvers.navstokes.kernels.flux'/>
+
+<% tau = c['ldg-tau'] %>
+
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur'>
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Is this necessary?

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%include file='pyfr.solvers.euler.kernels.rsolvers.${rsolver}'/>
+<%include file='pyfr.solvers.navstokes.kernels.bcs.common'/>
+<%include file='pyfr.solvers.navstokes.kernels.flux'/>
+
+<% tau = c['ldg-tau'] %>
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Defined but never used.  Unsure if we should be including a penalty term or not.

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako (hunk 1 above)
Comment (iyer-arvind) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Left it for clarity, makes no difference to the implementation.

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako (hunk 2 above)
Comment (iyer-arvind) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Removed it

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%include file='pyfr.solvers.euler.kernels.rsolvers.${rsolver}'/>
+<%include file='pyfr.solvers.navstokes.kernels.bcs.common'/>
+<%include file='pyfr.solvers.navstokes.kernels.flux'/>
+
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur'>
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: End up with code duplication and expanding (effectively) the same thing twice in the generated source.  Probably best to remove.

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    ur[0] = ul[0];
+% for i in range(ndims):
+    ur[${i + 1}] = ul[${i + 1}] - 2*nor*nl[${i}];
+% endfor
+    ur[${nvars - 1}] = ul[${nvars - 1}];
+</%pyfr:macro>
+
+<%pyfr:macro name='bc_common_flux_state' params='ul, gradul, nl, magnl'>
+    // Ghost states
+    fpdtype_t ur[${nvars}], gradur[${ndims}][${nvars}];
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Variable defined but not used.

File: pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+<%include file='pyfr.solvers.euler.kernels.rsolvers.${rsolver}'/>
+<%include file='pyfr.solvers.navstokes.kernels.bcs.common'/>
+<%include file='pyfr.solvers.navstokes.kernels.flux'/>
+
+<%pyfr:macro name='bc_common_flux_state' params='ul, gradul, nl, magnl'>
+    // Ghost state r
+    fpdtype_t ur[${nvars}];
+
+    fpdtype_t nor = ${' + '.join('ul[{1}]*nl[{0}]'.format(i, i + 1)
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Have you tested this?  If I recall you have to defined `bc_ldg_state` as we need it for the `bcconu` kernel.

Comment (iyer-arvind) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: I am testing it, push to github in this period is only for my comfort for shifting of code across machines. I will let you know after its duly tested.

Comment (iyer-arvind) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: I can see bcconu needs bc_ldg_state, but the code is still running(perhaps with erroneous results (or not!), I will need time to make that claim).  Just to figure out where is bcconu getting called?

Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/slp-adia-wall.mako]: Get PyFR to dump the expanded source code for all of the kernels.  Should quickly be able to see what is going on.

**Comment by iyer-arvind** [comment]
Tested to be working.

**Pull request merged**